### DISCUSSION TITLE
Ability to auto resolve correct option and label for value prop

### DISCUSSION THREAD

**Discussion by ScottMRafferty** [JedWatson/react-select#4829] [Ideas]
Would it be possible to resolve the correct labels for options when the value options are not normalised i.e. pass in {value: 1, label: 1} or not objects (value or array of values) i.e. 1 and have react-select auto resolve the correct option based on the internal options available.  I've experimented with the following 2 custom SingleValue and MultiValue components in a sandbox to do this:

```javascript
const SingleValue = (props) => {
  let valArr = props.getValue();
  if (props.options)
    valArr = valArr.map(valObj => props.options.filter((e) => e && e.value === valObj.value)[0] || valObj);
  return <components.SingleValue {...props}>{valArr.map(e=>e&&e.label)}</components.SingleValue>;
};

const MultiValue = (props) => {
  let valArr = props.getValue();
  if (props.options)
    valArr = valArr.map(valObj => props.options.filter((e) => e && e.value === valObj.value)[0] || valObj);
  return <components.MultiValue {...props}>{valArr.map(e=>e&&e.label)[props.index]}</components.MultiValue>;
};
```

It would be great to either have it automatically do this or at least have an autoResolveLabels prop to enable it.  

I also notice that the components are not fired if the default value is a value and not an object.  In the cases whereby I only have the values I can convert them to objects with the same value and label before I pass in but it would be great if I could just pass in a raw value or an array of raw values and react-select converted them internally to objects (with the label prop the same as the value prop) and then have the label autoresolve as above.

Some of the reasoning:

1. Occasionally it's useful to return the value(s) only from react-select and also have react-select take a value/values back in and find and display the correct label for the value(s) from the options.  At the moment I have to resolve the value and label before passing back in to react-select (or at the very least create an object with value and label).  When the component is in uncontrolled mode it's only an issue for the initial value but in controlled mode the value could be changed elsewhere and passed back in to react-select where the label just blanks out even if there is a valid option value that matches.
2. As react-select holds the list of options (granted not in async non-prefetch mode) then it would just offer up more flexibility in what can be passed in as values.  If values can be resolved to the correct options then why not do it?
3. I think this functionality would merge well with non-normalised form values that are fetched from backend with the options fetched from the backend.  Normalisation can then occur in the component.  I understand that I can normalise everything in advance of initialising react-select but I believe this flexibility would be powerful - at least I have compelling use cases and I'm sure others must have as well.

Just to be clear when I use the term normalise here I'm talking about taking a value ID and associating it with it's options object and ultimately normalising the ID to the label that the ID represents.

There is always the caveat that the label cannot be resolved if the options are aysnc fetched real time but as far as I can see in all other scenarios including async option prefetch the labels can be resolved if only a value or array of values are passed in.  In the case of real time async existing behaviour would continue as is.

Hope this makes sense.  If it seems like it might be useful I'm happy to make the changes and create a pull request but I thought it best to open up the discussion first.

Cheers.

**Comment by ebonow** [comment]
@ScottMRafferty Just so I am clear, is the request to allow options that are a String or Number type instead of an object or am I misinterpreting your request? 

Also, I didn't see anything mentioned about the props `getOptionLabel` or `getOptionValue` which take an object and resolves the label and value without needing to include them explicitly in the option data, so I just wanted to make sure you knew these were an option in case this was something you might have been looking for.

**Reply by ScottMRafferty** [reply]
Thanks for responding.  

Say I have the following options: 
```javascript
[{value: 1, label: 'red'), {value: 2, label: 'green'}, {value: 3, label: 'blue'}]
```
Passing in value prop:

Scenario 1: ```{value: 1, label: 'red'}```  : works fine and the label red displays.
Scenario 2: ```1``` : react-select does not display a label or fire the custom component label renderers.
Scenario 3: ```{value: 1, label: 1}``` : react-select by default will display 1 as the label but this does fire the custom component renderers so I can "patch it" using the custom components in my original post to pull the correct label from any existing options held internally by react-select (preset or aysnc prefetched). 

You can extrapolate the above for multi-values as well i.e. desired behaviour for passing in array of values [1,2,3] would have react-select presenting Red, Green and Blue for the labels.

Users can obviously map the value prop to the correct options before it reaches react-select but when using the async component  options are fetched AFTER the component has been initialised with the value. A hook could possibly be added before the options callback but that would get messy real fast I believe.

Not sure that getOptionLabel or getOptionValue help in this case as the options that are stored internally (and retrieved) by react-select are not passed over to those functions and the labels cannot be resolved without those options? I'm also not sure those functions even fire in scenario 2 i.e. raw value being passed in rather than an object but I'll double check that.

To sum up more technically in scenario 2 above react-select should be able to convert the raw value into a value, label object internally as it's value state i.e. 1 to {value:1, label:1} and then for all internal value state it should ATTEMPT to resolve/correct the object label prop against any internal options that it holds.  If there are no options that match the internal value state object value (!!!) then fall through to current behaviour.

Hope this clarifies things a little ;-)

**Reply by ebonow** [reply]
I think the biggest confusion is the usage of the word `value` both as the _selected option(s)_ and the _selected option(s) prop_. Conflating this difference could be problematic.

When a `value` is passed to the Select as a prop, it is considered a "controlled input". This means that the value should represent the desired state of the input. Passing in a prop value of `1` is a direct conflict with what the state should be as it expects an object, array of objects, null, or undefined (uncontrolled). Changing this paradigm so that the component should at some point change the value passed to represent something else sometimes is an anti-pattern. It's changing the input from being **controlled** to **uncontrolled** as the `value` prop no longer represents the selected option(s). 

Again, I believe this is all semantics caused by the naming structure of the prop called `value` where it may have been better named as `selected` whereas the value could simply have been the value of the selected option(s) (similar to an HTML Select element). In an HTML select, you can set the value, and it knows to highlight that option.

Mapping as you mentioned is possible as is the ability to do it async prior to the callback. Personally, I am accessing the internal methods on the selectRef to `setValue` when the `options` or a custom defined prop called `defaultSelectedOptionValue` changes as I wrote most of these prior to hooks. 

Regarding Async though, I think it's hard to set a default as it is essentially a wrapper for an autocomplete api and it's likely the selected option will not be in the options list.

Writing this in a hook would likely be straight forward enough and would be happy to do so, but at the end of the day, it is mapping back from the options and setting those as the value, so at some level it is still a controlled input that is likely waiting on the results of some api call that the dev is calling to then map and find the options/value, so it's not clear how beneficial this would/could be. 

I'd be happy to look at a codesandbox if you wanted to craft a proposal that didn't involve reusing the value prop for this.

**Comment by ScottMRafferty** [comment]
I agree that it could be considered an anti-pattern but the behaviour is definitely clashing with other legacy select components including react-select-plus (the previous incarnation).

Check out this sandbox.  I've lowered React to 15.4 as the older select components rely on prop types etc. but for users with legacy projects looking to just pop in react-select as a replacement the fact it can't resolve raw values to labels is a problem especially if source of truth for value state is held outside of the select components i.e. controlled by a form manager etc.

https://codesandbox.io/s/react-select-value-prop-behaviour-compared-nveoz

I have two large projects that I upgraded from react-select-plus to react-select and this has proven to be the biggest headache as all the form schemas and controllers are based on holding values as raw values i.e. lang="en" or client_id = 345 and I instruct the select component to only output the value for the selected option back to the form controller in onChange.  The form data values now have to be normalised in advance of passing into react-select (if async this is a pain) or normalised just before component instantiation (in the case of the options being available already).  As I have 70+ JSON form schemas normalising all the necessary form value props to their option equivalents has required quite a lot of code rework. I can kind of get around some of this now using the SingleValue and MultiValue custom components.

As far as type ahead async is concerned I agree with you that the raw value must be converted to the correct value object before passing in to react-select as there is no sensible way to derive the value object internally.  In all other cases I believe the internal state value object can be derived from the available options and fallback to existing behaviour if not.

I'm not saying that react-select should absolutely implement this behaviour (we have to move forward right?) but to my mind the component does deviate slightly from expected behaviour as far as it's own legacy component and other legacy components.  It's definitely cost me at least 50-60 hours of work to put right in legacy projects and I know I was quite suprised by the behaviour when I first switched to react-select some time back.  This may be due to the legacy components implementing bad patterns but a lot of legacy wrapper code will be written with that behaviour in mind.  I can't imagine that I'm the only one to have come across this BUT... :-)

In all react-select is an amazing component which is why I have perservered with this.  I hope the sandbox example provides some further context as it is probably the most basic example of the behavioural difference from 2 previous components to react-select.  You are also right that some of the reference terms (values, value objects, value object value) are confusing!!!

Cheers.
